### PR TITLE
disable 'fsync' by default, but force fsync() in more situations

### PR DESCRIPTION
**Pull request by justinmk**
fsync() is very slow on my system. It randomly causes `:write` and `:quit` to take 3+ seconds.

This PR does several things:

- Make shada and swapfile respect the 'fsync' option.
- Force `fsync()` (regardless of 'fsync' option) in these cases:
  - Idle (CursorHold).
  - Exit caused by deadly signal.
  - SIGPWR signal.
  - Explicit :preserve command.
- Disable the 'fsync' option by default.

Also mitigates #6725.

ref https://github.com/ludovicchabant/vim-gutentags/issues/167

### CODE REVIEW

**Comment by jamessan** [comment]
Just to clarify, this change means that when `:w` is issued, the file will be flushed after `'updatecount'` characters are typed?  How does that affects things like https://github.com/tpope/tpope/blob/e8f5d80cf46df8b962cd947a6575bd0621fd6da6/.vimrc#L329-L330 ?

The save would cause `'swapfile'` to be unset, but the file hasn't been `fsync()`d yet.

File: runtime/doc/options.txt [hunk 1]
```diff
@@ -2690,17 +2690,19 @@ A jump table for the options with a short description can be found at |Q_op|.
...
+'fsync' 'fs'		boolean	(default off)
+			global
+	When on, the OS function fsync() will be called after saving a file
+	(|:write|, |writefile()|, …), |swap-file| and |shada-file|. This
+	flushes the file to disk, ensuring that it is safely written.
+	Slow on some systems: writing buffers, quitting Nvim, and other
+	operations may sometimes take a few seconds.
+
+	Files are ALWAYS flushed ('fsync' is ignored) when:
+	- |CursorHold| event is triggered
```
[the remark this answers is not in the document]
Comment (justinmk) [reply runtime/doc/options.txt]: No it's 'updatetime'. `before_blocking()` calls `updatescript(0)` which is special-cased.

**Comment by justinmk** [comment]
> when :w is issued, the file will be flushed after 'updatecount' characters are typed?

The file itself is not fsync'd, only its "memfile" is, and only if the buffer is `&modified`. 

It would be nice to fsync the file asynchronously or after idle, but (I would guess) that involves managing file descriptor lifetimes. (And that would be my preference, so then 'fsync' can be enabled by default.)

My assumption is that swapfiles are enough to protect against data loss (within 'updatetime'). Also [this](http://blog.httrack.com/blog/2013/11/15/everything-you-always-wanted-to-know-about-fsync/) suggests that fsync() by itself is not always effective (maybe libuv takes extra measures?).

I left 9139bf81cf81a93e331f98dcddbe489fc3529787 as a separate commit if we want to revert it. But again, the goal should be to fsync asynchronously. #6725 also needs to be fixed.

references:

- https://thunk.org/tytso/blog/2009/03/15/dont-fear-the-fsync/ (incl. comments)
- https://unix.stackexchange.com/a/384450/

File: runtime/doc/options.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (justinmk) [reply runtime/doc/options.txt]: After reflection, I think it should also happen for 'updatecount'.

**Comment by justinmk** [comment]
>> The file itself is not fsync'd, only its "memfile" is, and only if the buffer is &modified.
>
> Now the file being saved is not fsync'd on write and the memfile is only sync'd after a delay.

Yes, if 'fsync' is not set. That's what I said, right? :)

> Or, as tytso suggested in your reference, do the write and fsync async (together).

Seems higher risk.  

- write-now-without-fsync only risks data loss if the OS crashes. 
- delayed-write-and-fsync risks data loss if the application crashes.

Different idea: what if we called `sync()` on 'updatetime' (or some other "idle" timer)?

**Comment by lewis6991** [comment]
I'm quite confused by this change.

@justinmk is `fsync` still a problem for you?

**Comment by justinmk** [comment]
@lewis6991 it depends on the filesystem. This PR isn't perfect, but in general we shouldn't need 'fsync' if we fsync swapfiles at 'updatetime'.

**Comment by lewis6991** [comment]
The default value of updatetime is 4 seconds though? That seems quite a lot.

**Pull request merged**